flax/errors.py

```diff
@@ -297,5 +323,263 @@ def __call__(self, x):
+
+    class Foo(nn.Module):
+      def setup(self):
+        self.dense = nn.Dense(features=4)
+
+      def __call__(self, x):
+        return self.dense(x)
+    
+    foo = Foo()
+    foo(np.arange(5))
+
+  The reason is that by directly calling ``foo(np.arange(5))``, ``setup()`` is
+  not called first, so ``self.dense`` is undefined. To resolve this, the module
+  should be called using :meth:`Module.apply() <flax.linen.Module.apply>` 
+  instead, which will internally try to call ``setup``::
+
+    foo = Foo()
+    variables = foo.init(rand.PRNGKey(0), np.arange(5))
+    foo.apply(variables, np.arange(5))
+  """
+  def __init__(self, module_name, attr_name):
+    super().__init__(f'"{module_name}" object has no attribute "{attr_name}"')
+
+
+class SetAttributeInModuleSetupError(FlaxError):
+  """
+  You are not allowed to modify Module class attributes in
+  :meth:`Module.setup() <flax.linen.Module.setup>`::
+
+    class Foo(nn.Module):
+      features: int = 6
+
+      def setup(self):
+        self.features = 3  # <-- ERROR
+
+      def __call__(self, x):
+        return nn.Dense(self.features)(x)
+
+    variables = SomeModule().init(random.PRNGKey(0), jnp.ones((1, )))
+
+  Instead, these attributes should be set when initializing the Module::
+
+    class Foo(nn.Module):
+      features: int = 6
+
+      @nn.compact
+      def __call__(self, x):
+        return nn.Dense(self.features)(x)
+
+    variables = SomeModule(features=3).init(random.PRNGKey(0), jnp.ones((1, )))
+        
+  """
+  def __init__(self):
+    super().__init__(f'Module construction attributes are frozen.')
+
+
+class SetAttributeFrozenModuleError(FlaxError):
+  """
+  You can only assign Module attributes to ``self`` inside
+  :meth:`Module.setup() <flax.linen.Module.setup>`. Outside of that method, the
+  Module instance is frozen (i.e., immutable). This behavior is similar to
+  frozen Python dataclasses.
+  
+  For instance, this error is raised in the following case::
+
+    class SomeModule(nn.Module):
+      @nn.compact
+      def __call__(self, x, num_features=10):
+        self.num_features = num_features  # <-- ERROR!
+        x = nn.Dense(self.num_features)(x)
+        return x
+
+    s = SomeModule().init(random.PRNGKey(0), jnp.ones((5, 5)))
+
+  Similarly, the error is raised when trying to modify a submodule's attributes
+  after constructing it, even if this is done in the ``setup()`` method of the
+  parent module::
+
+    class Foo(nn.Module):
+        def setup(self):
+          self.dense = nn.Dense(features=10)
+          self.dense.features = 20  # <--- This is not allowed
+        
+        def __call__(self, x):
+          return self.dense(x)
+  """
+  def __init__(self, module_cls, attr_name, attr_val):
+    super().__init__(f'Can\'t set {attr_name}={attr_val} for Module of type '
+                    f'{module_cls}: Module instance is frozen outside of '
+                     'setup method.')
+
+
+class MultipleMethodsCompactError(FlaxError):
```

Comment:
Also here would be nice as a TODO to add link to a short design note explaining this as it's key to having no need for an equivalent to hk.transparent, and makes submodule naming much more sane (no need to prefix the method name)